lib/puppet/provider/package/rpm.rb

```diff
@@ -155,22 +155,26 @@ def install
   end
 
   def uninstall
-    query if get(:arch) == :absent
-    nvr = "#{get(:name)}-#{get(:version)}-#{get(:release)}"
-    arch = ".#{get(:arch)}"
-    # If they specified an arch in the manifest, erase that Otherwise,
-    # erase the arch we got back from the query. If multiple arches are
-    # installed and only the package name is specified (without the
-    # arch), this will uninstall all of them on successive runs of the
-    # client, one after the other
-
-    # version of RPM prior to 4.2.1 can't accept the architecture as
-    # part of the package name.
+    query
+    # If a version and release is specified in the resource, uninstall that,
+    # otherwise do not specify a version and release to uninstall.
+    name    = get(:name)
+    version = get(:version)
+    release = get(:release)  
+    nvr     = "#{name}-#{version}-#{release}"
+    unless @resource[:name].start_with? nvr    
+      nvr = name
```

Comment:
It looks like if a user specified the resource name as `#{name}-#{version}` (without the release), we'd hit this case and drop the version information. That seems like it could be bad (uninstalling a different version of the package). Should this logic be expanded to handle that case?

Reply:
Thanks, I will take a look ...

Reply:
This (and removing the `unless` block) would handle that:

```
nvr = [get(:name), get(:version), get(:release)].join('-')
```

But the provider cannot query when `version` is specified but `release` and `arch` are not specified:

```
[root@pe-201902-master ~]# rpm -q vsftpd
vsftpd-3.0.2-25.el7.x86_64

[root@pe-201902-master ~]# rpm -q vsftpd-3.0.2-25.el7
vsftpd-3.0.2-25.el7.x86_64

[root@pe-201902-master ~]# rpm -q vsftpd-3.0.2
vsftpd-3.0.2-25.el7.x86_64

[root@pe-201902-master ~]# rpm -q vsftpd-3.0.2-25
package vsftpd-3.0.2-25 is not installed
```

Resulting in:

```
Debug: Prefetching yum resources for package
Debug: Executing '/bin/rpm -qa --nosignature --nodigest --qf '%{NAME} %|EPOCH?{%{EPOCH}}:{0}| %{VERSION} %{RELEASE} %{ARCH}\n''
Debug: Executing: '/bin/rpm -q vsftpd-3.0.2-25 --nosignature --nodigest --qf '%{NAME} %|EPOCH?{%{EPOCH}}:{0}| %{VERSION} %{RELEASE} %{ARCH}\n''
Debug: Executing: '/bin/rpm -q vsftpd-3.0.2-25 --nosignature --nodigest --qf '%{NAME} %|EPOCH?{%{EPOCH}}:{0}| %{VERSION} %{RELEASE} %{ARCH}\n' --whatprovides'
Debug: /Stage[main]/Main/Package[vsftpd-3.0.2-25]: Nothing to manage: no ensure and the resource doesn't exist
```

So I think I need to strip `release` from `@resource[:name]` in `query` when `arch` is not specified.

Reply:
Yes, addressed.